include/tvm/arith/linear_system.h

```diff
@@ -0,0 +1,171 @@
+ * \sa LinearSystemNode
+ */
+class LinearSystem : public ObjectRef {
+ public:
+  /*!
+   * \brief Constructor by fields
+   * \param variables The variables in the system.
+   * \param ranges    The ranges of the variables.
+   * \param relations The linear relations between the variables
+   *                  (either equations or inequalities)
+   */
+  TVM_DLL LinearSystem(Array<Var> variables,
+                       Map<Var, Range> ranges,
+                       Array<PrimExpr> relations);
+
+  TVM_DEFINE_OBJECT_REF_METHODS(LinearSystem, ObjectRef, LinearSystemNode);
+};
+
+/*!
+ * \brief We can have different set of variables to represent the same linear system.
+ *        For example, the following two systems are equivalent,
+ *        {a + b = 0 | a >= 0, b >= 0} and
+ *        {m - n = 0 | m >= 0, n <= 0}
+ *        This data structure represents the transformation
+ *        between two equivalent linear systems.
+ *        In the above example,
+ *        src        : {a + b = 0 | a >= 0, b >= 0}
+ *        dst        : {m - n = 0 | m >= 0, n <= 0}
+ *        src_to_dst : {a -> m, b -> -n}
+ *        dst_to_src : {m -> a, n -> -b}
+ * \sa LinearSystemTransform
+ */
+class LinearSystemTransformNode : public Object {
+ public:
+  LinearSystem src;
+  LinearSystem dst;
+  Map<Var, PrimExpr> src_to_dst;
+  Map<Var, PrimExpr> dst_to_src;
+
+  void VisitAttrs(tvm::AttrVisitor* v) {
+    v->Visit("src", &src);
+    v->Visit("dst", &dst);
+    v->Visit("src_to_dst", &src_to_dst);
+    v->Visit("dst_to_src", &dst_to_src);
+  }
+
+  static constexpr const char* _type_key = "arith.LinearSystemTransform";
+  TVM_DECLARE_FINAL_OBJECT_INFO(LinearSystemTransformNode, Object);
+};
+
+/*!
+ * \brief Managed reference to LinearSystemTransformNode.
+ * \sa LinearSystemTransformNode
+ */
+class LinearSystemTransform : public ObjectRef {
+ public:
+  /*!
+   * \brief Constructor by fields
+   * \param src        source linear system, e.g., {a + b = 0 | a >= 0, b >= 0}
+   * \param dst        linear system equivalent to the source, e.g., {m - n = 0 | m >= 0, n <= 0}
+   * \param src_to_dst mapping from variables in the \p src to the variables in the \p dst,
+   *                   e.g., {a -> m, b -> -n}
+   * \param dst_to_src mapping from variables in the \p dst to the variables in the \p src,
+   *                   e.g., {m -> a, n -> -b}
+   */
+  TVM_DLL LinearSystemTransform(LinearSystem src,
+                                LinearSystem dst,
+                                Map<Var, PrimExpr> src_to_dst,
+                                Map<Var, PrimExpr> dst_to_src);
+
+  TVM_DEFINE_OBJECT_REF_METHODS(LinearSystemTransform, ObjectRef, LinearSystemTransformNode);
+};
+
+/*!
+ * \brief Obtain Smith Normal Form of linear equation A x = y.
+ *        Smith Normal Form of matrix A_{mxn} is S_{mxn} = U_{mxm} A_{mxn} V_{nxn},
+ *        in which S_{mxn} is diag(s1, s2, ..., sr, 0, ..., 0),
+ *        such that si | s_{i+1} and r is the rank of A.
```

Comment:
I'm not sure, but I think the condition `si | s_{i+1}` is not guaranteed by this implementation, so it's not fully fledged Smith Normal Form (which shouldn't be a problem for our end goal).